SnM/SnM_Notes.cpp

```diff
@@ -1472,17 +1472,29 @@ void WriteGlobalNotesToFile()
 	// write global notes to <REAPER resource path>/SWS_Global notes.txt
 	WDL_FastString filePath;
 	filePath.SetFormatted(SNM_MAX_PATH, "%s/SWS_Global notes.txt", GetResourcePath());
-	WDL_FileWrite outfile(filePath.Get(), 1, 65536, 16, 16); // NF: not sure about these params, taken from 
-	//https://github.com/justinfrankel/licecap/blob/3721ce33ac72ff05ef89d2e92ca58a0f96164134/WDL/lice/lice_gif_write.cpp#L386
-	int wrtPos = outfile.Write(g_globalNotes.get_filepart(), g_globalNotes.GetLength());
-	if (wrtPos == g_globalNotes.GetLength()) // writing OK
+
+	if (g_globalNotes.GetLength())
 	{
-		g_IsGlobalNotesDirty = false;
+		WDL_FileWrite outfile(filePath.Get(), 1, 65536, 16, 16); // NF: not sure about these params, taken from 
+		//https://github.com/justinfrankel/licecap/blob/3721ce33ac72ff05ef89d2e92ca58a0f96164134/WDL/lice/lice_gif_write.cpp#L386
+		int wrtPos = outfile.Write(g_globalNotes.get_filepart(), g_globalNotes.GetLength());
+		if (wrtPos == g_globalNotes.GetLength()) // writing OK
+		{
+			g_IsGlobalNotesDirty = false;
+		}
+		else // writing failed
+		{
+			MessageBox(GetMainHwnd(), __LOCALIZE("Saving global notes failed.", "sws_mbox"), __LOCALIZE("SWS - Error", "sws_mbox"), MB_OK);
+		}
 	}
-	else // writing failed
+	else // remove if empty
 	{
-		MessageBox(GetMainHwnd(), __LOCALIZE("Saving global notes failed.", "sws_mbox"), __LOCALIZE("SWS - Error", "sws_mbox"), MB_OK);
-	}
+		std::remove(filePath.Get());
```

Comment:
Mhh I'm pretty sure this fails on Windows if the file path contains characters outside of the current system ANSI codepage.

If the file must be removed when empty, using the wide version of `remove` should do the trick:
```suggestion
#ifdef _WIN32
		_wremove(win32::widen(filePath.Get()).c_str());
#else
		remove(filePath.Get());
#endif
```

Reply:
Thanks for reviewing.
> If the file must be removed instead of saved empty
I'm not very confident with it myself (think empty text file is not that much harm), but it's in reaction to [edgemeal's comment](https://forum.cockos.com/showpost.php?p=2233645&postcount=2755) the file appearing despite not using this feature. 

What I particularly don't like, since global notes are saved on project save, it now tries to delete this file every time, also when it's not there (user hasn't used the feature). I could check before if the file actually exists but I guess that's not more efficient than calling `remove()` directly, as this also checks the existence first anyway?
But then again there should be a way to handle when user purposely wants to save empty notes (to wipe previous ones) so I think I can't omit the `g_globalNotes.GetLength() == 0` condition completely.
Do you by chance have a better idea how to handle it?

Reply:
Found a better way (I think).
Checking for `g_IsGlobalNotesDirty` before saving prevents creating the SWS_Global Notes.txt file if user isn't using this feature (and no need for std::remove anymore).

Reply:
Good idea! (I was going to suggest using a smart pointer for `g_globalNotes` defaulting to null if the user didn't use global notes.)

Could the notes be saved when closing the notes window (eg. from `NotesWnd::OnDestroy`)? Is there a reason saving any project is required (or using the menu)?

Also I found a bug: because `WriteGlobalNotesTofile` uses `WDL_FastString::get_filepart` instead of `Get`, everything before (and including) the last slash is discarded and garbage from outside of the buffer is appended (since it's still reading `GetLength` worth of characters).

(nitpick: the extra indentation in `WriteGlobalNotesToFile` can be avoided by returning early: `if (!g_globalNotesDirty) return;`)

Reply:
Thanks!
>  I found a bug:

Should be fixed, just force-pushed.

> nitpick:
Also done.

> Could the notes be saved when closing the notes window (eg. from NotesWnd::OnDestroy)? Is there a reason saving any project is required (or using the menu)?

Why does Reaper provide a 'Save' menu item instead of just saving automatically when quitting? :P
More seriously, I actually regarded it as a feature providing a means to the user saving the global notes programmatically and not as a requirement (i.e. hassle).
Case :): user  is in progress writing his/her 10.000+ chars global notes, Reaper crashes while doing something else intermittedly, notes would be lost if not saved before (`NotesWnd::OnDestroy` wouldn't be called in this case).
And I find the most expected/user friendly way is that 'stuff' gets saved with a project save (I regard the saving from notes window window just as 'sugar on top').

Would that be a valid argument you think? :)

Reply:
> Would that be a valid argument you think? :)

Yes, thanks for explaining!

> Why can't I use `std::strlen` on Linux (and `strlen` works)?

`std::strlen` is defined in the `cstring` standard C++ header. SWS explicitly includes the C variant (`string.h`) which may or may not put functions in `std`. `cstring` is probably getting included by another header on macOS and Windows.

Reply:
Just read the new patch. The buffer over-read is fixed, but shouldn't it save the whole thing that the user typed rather than only what's after the last slash/backslash?

```cpp
const int wrtPos = outfile.Write(g_globalNotes.Get(), g_globalNotes.GetLength());
```

Reply:
Oh, somehow I got totally confused with the `get_filepart` function. I somehow thought I'd need it for saving the file, just looked at the definition now what it actually does. :1st_place_medal: 

So just:
```
const int wrtPos = outfile.Write(g_globalNotes.Get(), g_globalNotes.GetLength());
if (wrtPos == g_globalNotes.GetLength()) // writing OK
{ ...
```

right?
Thanks.

Reply:
Yes (either way is fine due to `WDL_FastString::GetLength`'s constant-time complexity. `globalNotesLength` should probably be an `int` to match the return type of `GetLength`).